airbyte-cdk/bulk/README.md

```diff
@@ -1,15 +1,107 @@
+As the name suggests, its purpose is to help develop connectors which extract or load data in bulk.
+The Bulk CDK is written in Kotlin and uses the Micronaut framework for dependency injection.
+
+## Structure
+
+The Bulk CDK consists of a _core_ and a bunch of _toolkits_.
+
+### Core
+
+The _core_ consists of the Micronaut entry point and other objects which are expected in
+connectors built using this CDK.
+
+The core is broken down into multiple gradle projects; for example the core functionality for
+building sources is in `extract`.
+
+Following up on that example, the expectation for a source connector is that it will use all the
+interfaces and implementations in `extract` unless it has a very good reason not to.
+There is plenty of value in having all source connectors behave predictably.
+
+### Toolkits
+
+The _toolkits_ consist of optional modules which contain objects which are common across
+multiple (but by no means all) connectors.
+
+For example, there's an `extract-jdbc` toolkit to help build source connectors which extract data
+using the JDBC API.
+The expectation for a toolkit is that it provides naive implementations of core interfaces.
+These implementations will be thoroughly tested inside the CDK to serve as a baseline of
+functionality; however the connector may (and in fact often should!) replace parts of these.
+
+Following up on the example of `extract-jdbc`, a source connector needs to implement SQL query
+generation interfaces and, for schema discovery, may prefer to query system tables directly
+instead of relying on the generic JDBC metadata methods.
+
+## Dependencies
+
+The Bulk CDK gradle build relies heavily on so-called [BOM dependencies](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#bill-of-materials-bom-poms).
+This pattern is strongly encouraged to keep transitive version conflicts to a minimum.
+
+Consider for example the whole Jackson ecosystem.
+Using a BOM allows us to add specific Jackson dependencies without having to figure out which
+version number to use.
+This has some pleasant ripple-effects:
+
+- When the need comes to bump the version, there's only one version number to bump and that's in
+  the BOM import.
+  Consequently, the declared version has a much higher chance of being the effective version
+  picked by gradle during dependency resolution.
+
+- The BOM import is re-exported by the `bulk-cdk-core-base` artifact meaning that the rest of the
+  CDK as well as connectors don't need to worry about Jackson version numbers either.
+
+It gets better when multiple BOMs are involve.
+Consider for example Micronaut and Jackson: Micronaut also depends on Jackson.
+This can (and will!) cause dependency version conflicts; these are much easier to resolve by
+reconciling just two BOM versions.
+
+While BOMs are undoubtedly useful, let's still try to keep external dependencies to a minimum
+outside of tests.
+Less dependencies, less problems.
```

Comment:
Is there some guidance that we want the lower parts of the CDK to have as few dependencies as possible, but the toolkits can gain deps?

Reply:
I don't know about that. Toolkits should also stay slim, in my opinion. Where it matters a lot less is tests, but even so, with gradle when you have a lot of dependencies (and therefore, conflicts) it's super easy to declare depending on version X but really because of a transitive dependency you're using version Y instead, without knowing.